Kinetics mirge generator

MTCam opened the pull request:
Here's a set of changes to the Pyrometheus code generator that I will be testing in _MIRGE-Com_.   These pass the original _pyrometheus_ tests in the `kinetics` branch.   

This branch is required by the [_MIRGE-Com_ Pyrometheus EOS](https://github.com/illinois-ceesd/mirgecom/pull/236).

MTCam commented:
I am happy to report that [mirgecom EOS and Chem tests ](https://github.com/illinois-ceesd/mirgecom/pull/164)are passing with this version of the generator.  The `sanDiego` mechanism is failing due to the treatment of the falloff functions which are not present in UIUC mech.

ecisneros8 commented:
@MTCam Did you branch from kinetics? I had falloff functions working (read: passing tests) there since quite a few commits ago. Look here: 

[](https://github.com/inducer/pyrometheus/blob/e3ff514f09e38db02b34da4b0b9cbd6530cb177c/pyrometheus/__init__.py#L235)

MTCam commented:
> @MTCam Did you branch from kinetics? I had falloff functions working (read: passing tests, for San Diego) there since quite a few commits ago. Look here:
> 
> [https://github.com/inducer/pyrometheus/blob/e3ff514f09e38db02b34da4b0b9cbd6530cb177c/pyrometheus/**init**.py#L235](falloff in kinetics branch)

Yes, the base branch of this PR is the kinetics branch in #4.  The generator in this branch still passes the native tests including `sanDiego` mechanism + falloff functions.  It is the MIRGE-specific multi-point test that is failing for `sanDiego` in MIRGE-Com only.

inducer commented:
I just merged master into #4. Please bring this up to date with respect to that branch, and add documentation. In addition, `npctx` is not a good name. Maybe `species_array_np` (or `species_np` or `spec_ary_np`) is a better name.

ecisneros8 commented:
Thanks @inducer I’m starting documentation today.

inducer commented on pyrometheus/__init__.py:
```diff
@@ -329,37 +330,84 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+    def __init__(self, actx_np=np):
+        \"""Initialize the mechanism object.
+
+        Parameters
+        ----------
+        actx_np
+            numpy-like namespace providing at least the following functions:
+            actx_np.log(T)
+            actx_np.log10(T)
+            actx_np.exp(T)
+            actx_np.where(T > 0, T_yes, T_no)
+            actx_np.linalg.norm(T, np.inf)
+            where T is model user data having the same size, shape, and nature
+            of the user's temperature data, for example. This parameter allows
+            the user to provide a numpy work-alike that is capable of processing
+            the user's data structures, or even invoking custom kernels for the
+            associated operations. This parameter defaults to *actual numpy*, so
+            unless it is needed by the user (e.g. for GPU processing).
```
Nobody sees this docstring. Please move this to the documentation of the `Thermochemistry` interface at the top of this file.

MTCam replied:
My attempts to move this documentation were not entirely successful.  I could not find any examples of how to document class member methods in a global docstring.   `make html` complains about my section headings.  I kludged it together and it currently makes an acceptably rendered (if incomplete) document.

inducer commented on pyrometheus/__init__.py:
```diff
@@ -329,37 +330,84 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+            the user to provide a numpy work-alike that is capable of processing
+            the user's data structures, or even invoking custom kernels for the
+            associated operations. This parameter defaults to *actual numpy*, so
+            unless it is needed by the user (e.g. for GPU processing).
```
This is a good start to document what is expect of `actx_np`, but it needs to be pulled out into a separate interface and actually make functional use of RST markup.

MTCam replied:
Plz see below

inducer commented on pyrometheus/__init__.py:
```diff
@@ -329,37 +330,84 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
-    num_reactions = ${sol.n_reactions}
-    num_falloff   = ${
-        sum(1 if isinstance(r, ct.FalloffReaction) else 0
+    def __init__(self, actx_np=np):
```
I maintain that `actx_np` is not acceptable as a name. It suggests that the only thing this is usable for is meshmode's array context. That's not true. Instead, we need to describe what it is used for, and what is expected of it. The documentation you provide below begins to define that interface.

MTCam replied:
I took a better stab at fixing this up including the naming and documentation.  I am hoping this is closer to what you are looking for.

MTCam replied:
62615a7

ecisneros8 commented on pyrometheus/__init__.py:
```diff
@@ -409,7 +435,7 @@ def get_concentrations(self, rho, Y):
         conctest = self.iwts * rho * Y
         zero = 0 * conctest[0]
         for i, conc in enumerate(conctest):
```
Where is conc used in this loop?

MTCam replied:
We *may* use it instead of `concs[index]`.  More testing is needed to make sure this flies.   I had trouble with it, and switched back to using indexed data structure in debugging.  I suggest we ignore this for now.

ecisneros8 commented on pyrometheus/__init__.py:
```diff
@@ -409,7 +435,7 @@ def get_concentrations(self, rho, Y):
         conctest = self.iwts * rho * Y
```
Why conctest? My issue with this variable name is the 'test' part. Can we call it concentrations or simply c?

ecisneros8 commented on pyrometheus/__init__.py:
```diff
@@ -868,42 +523,43 @@ def get_temperature(self, enthalpy_or_energy, t_guess, y, do_energy=False):
 
     def get_falloff_rates(self, T, C, k_fwd):
-        k_high = make_obj_array([
+        ones = (1.0 + T) - T
```
I've seen this is a couple of places now, and I'm just curious as to what it does? Is it a way of creating an array of 1's?

MTCam replied:
It's 1's, but importantly respecting the size, shape, and nature (let's say) of *T*, whatever *T* may be.

ecisneros8 commented on pyrometheus/__init__.py:
```diff
@@ -952,8 +609,8 @@ def get_net_rates_of_progress(self, T, C):
         r_net = self.get_net_rates_of_progress(T, C)
-        ones = r_net[0] + 1.0 - r_net[0]
-        return make_obj_array([
+        ones = (1.0 + r_net[0]) - r_net[0]
```
Analogous to my previous comment.

ecisneros8 commented on test/test_codegen.py:
```diff
@@ -30,6 +30,17 @@
+def test_generate_mechfile(mechname):
+    """This function tests that pyrometheus-generated code."""
+    sol = ct.Solution(f"{mechname}.cti", "gas")
+    mfile = open(f"{mechname}.py", "w")
```
I like src_file, py_file, or even mech_file instead. But I'll live.

ecisneros8 commented on test/test_codegen.py:
```diff
@@ -127,7 +143,9 @@ def test_get_thermo_properties(mechname):
         keq_err = np.abs((keq_pm - keq_ct) / keq_ct).max()
```
This propagated from kinetics, but my PR has a comment by @inducer to use `np.linalg.norm` instead.

MTCam replied:
The thing I absolutely hate about the np.linalg norm is that it processes the entire vector and erases any component-specific information.   Typically, when doing the numerical debugging - I like to see the error in each component separately because it can help guide one to the problem(s).  So far I have been unsuccessful at arguing for code that allows component-by-component inspection; so changed.

ecisneros8 commented on pyrometheus/__init__.py:
```diff
@@ -360,9 +358,37 @@ def _my_make_array(res_list):
+        ----------
+        usr_np
+            numpy-like namespace providing at least the following functions:
+            usr_np.log(T)
```
It is difficult to understand that this is not uniquely tied to the temperature. Can we either (i) state we're using T and temperature as an example up front; or (ii) generalize and then provide T and temperature as an example? (Hoping this is clear, otherwise let me know.)

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+    equiv_ratio = 1.0
+    ox_di_ratio = 0.21
+    stoich_ratio = 3.0  # 0.5
+    i_fu = sol.species_index(f"{fuel}")
```
```suggestion
    i_fu = sol.species_index(fuel)
```

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+    init_temperature = 1500.0
+    equiv_ratio = 1.0
+    ox_di_ratio = 0.21
+    stoich_ratio = 3.0  # 0.5
```
Previously, this depended on the fuel, but it no longer does. Why?

MTCam replied:
This setting is of no consequence to the actual test.   We are testing against Cantera - so as long as the Cantera state is specified identically to the Pyrometheus state, the test is valid.   Regardless, I switched this back to a fuel-dependent stoichiometric ratio.

MTCam replied:
642cfe6

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+    x[i_fu] = (ox_di_ratio*equiv_ratio)/(stoich_ratio+ox_di_ratio*equiv_ratio)
+    x[i_ox] = stoich_ratio*x[i_fu]/equiv_ratio
+    x[i_di] = (1.0-ox_di_ratio)*x[i_ox]/ox_di_ratio
+    sol.TPX = init_temperature, ct.one_atm, x
```
Is this used anywhere?

MTCam replied:
Yes, it is used on the very next line to initialize the reactor initial state.

MTCam replied:
Just to clarify (sol.TPX = ) sets an internal state for the _Cantera_ `sol` object.  Subsequent uses of the `sol` object depend on this data.

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+        r_pm = ptk.get_net_rates_of_progress(temp, c)
+        omega_pm = ptk.get_net_production_rates(rho, temp, y)
+        # Print
+        err_r = np.linalg.norm((r_ct-r_pm), np.inf)
```
```suggestion
        err_r = np.linalg.norm(r_ct-r_pm, np.inf)
```

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+        omega_pm = ptk.get_net_production_rates(rho, temp, y)
+        # Print
+        err_r = np.linalg.norm((r_ct-r_pm), np.inf)
+        err_omega = np.linalg.norm((omega_ct[0:-1]-omega_pm[0:-1]), np.inf)
```
Comment on why the last element of the array is being left out.

MTCam replied:
I don't know.  Alternative answer: these are not the array elements you are looking for. :jedi-hand-wave:

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+
+    time = 0.0
+    for step in range(50):
+        time += 1.0e-7  # 1.0e-6 for H2
```
I'm confused by the H2 comment here. Could you explain?

MTCam replied:
Just a note to self(s). Ethylene fuel needed 1e-7 to be stable.   1e-6 was sufficient for H2 fuel.

MTCam replied:
Fixed this up in (6ea96c8)

inducer commented on test/test_codegen.py:
```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
-    computes thermodynamic properties c_p, s_R, h_RT, and k_eq
-    correctly by comparing against Cantera"""
-    # Create Cantera and pyrometheus objects
+    computes the right rates of progress for given temperature
+    and composition"""
```
Specify "right". (I know you didn't add this, but now is as good a time as any.)

MTCam replied:
I don't understand what you are asking for.

MTCam replied:
I took this to mean you wanted to clarify that "right" means.   In this case right = Cantera-predicted.

MTCam replied:
4b6c51b

inducer reviewed (commented):
Thanks for working on this. A few thoughts below.

inducer commented:
> Update pyrometheus/__init__.py 

@MTCam Could you try and enter sensible commit messages?

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -454,57 +514,51 @@ def get_temperature(self, H_or_E, T_guess, Y, do_energy=False):
+            j = -pv_fun(t_i, y)
+            dt = -f / j
+            t_i += dt
+            if _pyro_norm(self.usr_np, dt, np.inf) < tol:
```
There are a couple of things to consider that tie into this. An easy way to think about the root of the issue is that the output temperatures are sensitive to the initial guess passed to this routine.   The ideal situation is that we pass the actual/current temperature as the initial guess - we typically settle for a "reasonable" guess or for the temperature at the last time step.

If an initial guess is far off from the actual temperature, then this Newton iter can (a) take much more iters to converge to good temp, (b) converge to the wrong result or (c) fail to converge within num_iter. (a), (b) are the main things we want to avoid with restarting the DV (at least temperature for now).  Additionally, since the output temps are sensitive to input temps - we *need* to restart temperature in order to get deterministic results through restart - and this is important.

Unless we fix the temperature restart issues, we can expect that we will regularly encounter situation (c),  but because the starting guess and the actual temperature are not close.  In this case, we don't want to pass back NAN, we want to actually find the valid temperature, and so reverting to a more expensive procedure to find it would be a reasonable thing to do.   Or - until we get it fixed we can set num_iter to something insanely high.  Heh.

However if we *fix* the temperature restart issue and start actually using the current(ish) temperature as the initial guess, then if we could stop the code on encountering condition (c), it'd be terrific.  

Another potential fix is to roll a routine that provides a deterministic guess for a starting temperature based on the input state.   Then we'd only need to save state as we strongly desire, and we'd still get a deterministic temperature calc.   The big problem there is coming up with a sane temperature estimator routine; but it might be relatively easy to roll mechanism=specific estimators right into the generated code.

inducer replied:
> An easy way to think about the root of the issue is that the output temperatures are sensitive to the initial guess passed to this routine.

Just so I have this clear: What you're saying is that the equation being solved here doesn't suffice to fully define the temperature, i.e. the  initial guess matters? (If so, that seems strange to me.)

> Unless we fix the temperature restart issues, we can expect that we will regularly encounter situation (c), but because the starting guess and the actual temperature are not close.

Sure, Newton can fail. But wouldn't you rather know that it did than not?

A different aspect is what this dependency does to some of the interfaces we have in mirgecom. IIUC what you're saying, then you can't reasonably ask for temperatures without supplying a starting guess. That means the EOS interface in mirgecom needs to change.

(If you see this in time, could you add these discussions to the agenda for tomorrow?)

MTCam commented on test/test_codegen.py:
```diff
@@ -30,56 +30,15 @@
-        assert np.linalg.norm(omega_ct-omega_pm, np.inf) < 1e-10
-
-    return
+    mech_file = open(f"{mechname}.py", "w")
+    code = pyro.gen_thermochem_code(sol)
+    print(code, file=mech_file)
+    mech_file.close()
```
Got it.

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -329,37 +349,64 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+    return result
+
+
+def _pyro_norm(usr_np, argument, normord):
+    # Wrap norm for scalars
+    from numbers import Number
+    if isinstance(argument, Number):
+        return np.abs(argument)
+    return usr_np.linalg.norm(argument, normord)
```
Could be.  This was your suggested fix after we discussed it some time back.

MTCam replied:
I scanned through pyopencl.Array interface and it wasn't apparent to me how I would fix this inside there.  I will start a PR for it if you can help guide my effort just a bit to where I should start.

MTCam replied:
@inducer added [this PR](https://github.com/inducer/meshmode/pull/100) to extend norm to numbers.  :tada:

Testing....

MTCam replied:
It turns out that the issue is that `numpy.linalg.norm(5, np.inf)` is not supported, so we need our wrapper after all.

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -329,37 +349,64 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
-        sum(1 if isinstance(r, ct.FalloffReaction) else 0
+    def __init__(self, usr_np=np):
+        self.usr_np = usr_np
+        self.model_name = ${repr(sol.source)}
+        self.num_elements = ${sol.n_elements}
+        self.num_species = ${sol.n_species}
+        self.num_reactions = ${sol.n_reactions}
+        self.num_falloff   = ${
+            sum(1 if isinstance(r, ct.FalloffReaction) else 0
```
Not sure I remember why I changed this.  I think it was when I added the argument for passed-in user numpy-like namespace.

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -384,40 +431,53 @@ def get_mix_molecular_weight(self, Y):
-        e0_RT = self.get_species_enthalpies_RT( T ) - 1.0
-        return self.gas_constant * T * np.sum(Y * e0_RT * self.iwts)
+        concs = self.iwts * rho * Y
+        zero = 0 * concs[0]
+        for i, conc in enumerate(concs):
+            concs[i] = self.usr_np.where(concs[i] > 0, concs[i], zero)
+        return concs
```
That does not work as intended:
```py
E   ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

MTCam replied:
Ah, so I still need to iterate through the array to do that?   If this expression is correct, then this "works" for `pyrometheus` proper - but unsure if it will continue working with the array context.

```
for (...):
   concs[i] = max(concs[i], 0)
```

MTCam replied:
Turns out, didn't work.  If this change is made, the code stops getting the right answers on all the points in the MIRGE-Com context as tested [here](https://github.com/illinois-ceesd/mirgecom/pull/147).

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -384,40 +431,53 @@ def get_mix_molecular_weight(self, Y):
+
+    def get_mixture_specific_heat_cp_mass(self, temperature, massfractions):
+        cp0_r = self.get_species_specific_heats_R(temperature)
+        cpsum = sum([massfractions[i] * cp0_r[i] * self.iwts[i]
+                     for i in range(self.num_species)])
```
This pattern appears several times, but for different properties.  Regardless, I refactored this a bit.  See how it grabs you now.

MTCam replied:
6a644b6

MTCam commented:
Scanning back through all the comments, I believe (most) everything has been addressed.  There is an open question about how to handle the specialized norm function `_pyro_norm`, but perhaps we can address that in an issue or separate PR?

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -319,6 +338,7 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
     """:returns: Species production rate for species *species* in terms of
     the net reaction rates of progress *r_net* as a
     :class:`pymbolic>primitives.Expression`"""
+    ones = (r_net[0] + 1.0) - r_net[0]
```
Yeah, this is very hokey code.  I added a `_zeros_like` to try to clean it up a bit, but I suspect once you see what I did there that we'll need to talk about it a bit further.

MTCam commented:
A couple of things:
 - rolled in the flake8 CI for generated code from #12 
 - updated the script to generate and flake8 *all* available mechs 
 - moved the mech input files to test/mechs/{mechname}.cti
 - moved generated mech output to test/mechs/{mechname}_mech.py

It might be nicer to have the `mechs` directory with the CTI files live at top level instead of down in tests.   Other than that I am happy with how this is working.

inducer commented on .ci/generate-test-mech.py:
```diff
@@ -0,0 +1,10 @@
+for mech in test_mechs:
+    mechname = mech[11:-4]
+    with open(f"test/mechs/{mechname}_mech.py", "w") as outf:
+        code = pyro.gen_thermochem_code(ct.Solution(f"{mech}", "gas"))
+        print(code, file=outf)
```
```suggestion
        print(pyro.gen_thermochem_code(ct.Solution(f"{mech}", "gas")), file=outf)
```

inducer commented on .ci/generate-test-mech.py:
```diff
@@ -0,0 +1,10 @@
+test_mechs = glob.glob("test/mechs/*.cti")
+for mech in test_mechs:
+    mechname = mech[11:-4]
+    with open(f"test/mechs/{mechname}_mech.py", "w") as outf:
```
The sphinx bits in #12 rely on a mechanism appearing at `pyrometheus.thermochem_example`. (But I realize there are multiple mechanisms here. Maybe emit the first one into `pyrometheus.thermochem_example`, too?

MTCam replied:
Oh sorry, I did not see that.  I'll get that fixed up.

Longer term, it might be nice if we had some doc support for the general interface, and then a little blurb about each mechanism - maybe from embedded documentation, in which case we'd want to have all of them represented in actual py code?

MTCam replied:
I refactored it just a bit with a function that generates a specified mech py file from a specified cti file. See the utils in `.ci` and `.github/workflows` yml file `doc` action for the generation of the required example.

MTCam replied:
ab15da3, 68a9507

MTCam commented:
> @MTCam Could you try and enter sensible commit messages?

Of course!  I intended to do that.  I didn't realize the bits that I wrote were showing up only in the dropdown until it was too late.

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -329,123 +328,228 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+            concs[i] = self.usr_np.where(concs[i] > 0, concs[i], zero)
+        return concs
+
+    def sum_over_species(self, mass_fractions, proparray):
```
Just need the function to be a "verb".   Try this on for size. 
8dda1a7

inducer commented on test/test_codegen.py:
```diff
@@ -128,26 +88,84 @@ def test_get_pressure(mechname):
+        print(f"keq_pm = {keq_pm}")
+        print(f"keq_cnt = {keq_ct}")
+        print(f"temperature = {t}")
+        # Exclude check on irreversible reactions
```
Why? Maybe add some context on why that's difficult.

MTCam replied:
Pyro doesn't "do" irreversible reactions.  Help me out @ecisneros8 ?

ecisneros8 replied:
Pyro does do irreversible reactions. We even compute their equilibrium constants, even though they are meaningless and never factor into the computation of the rates, to maintain the equilibrium-constants array at size N_reactions. However, we don't compare them against Cantera because, well, they are meaningless.

MTCam replied:
So my current plan would be to replace the comment with:
`# Exclude meaningless check on equilibrium constants for irreversible reactions`

inducer replied:
```suggestion
        # Exclude meaningless check on equilibrium constants for irreversible reactions
```

inducer reviewed (commented):
Sorry about taking forever on this. Here are some thoughts on this PR.

ecisneros8 commented on pyrometheus/__init__.py:
```diff
@@ -327,123 +326,232 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+        concs = self.iwts * rho * mass_fractions
+        zero = _pyro_zeros_like(concs[0])
+        for i, conc in enumerate(concs):
+            concs[i] = self.usr_np.where(concs[i] > 0, concs[i], zero)
```
I found that it is messing up Jacobians computed using JAX. All Jacobian entries associated with species with zero concentration come back zero, even when they shouldn't be. (Note, even when a mass fraction is zero, its corresponding Jacobian entries can be non-zero). I got rid of it, and I was able to verify the Jacobian of get_net_production_rates against finite differences and analytical expressions.

I'd much prefer these sorts of squashing to be done outside of pyro, and this is one example of why it's not best practice. Hope this is clear. Let me know otherwise.

MTCam replied:
So you allow negative concentrations in Pyro?

MTCam replied:
Can we try this: (3fcdd54)?

inducer replied:
Agree with @ecisneros8, we should probably do this outside.

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -327,123 +326,228 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+    .. method:: get_species_gibbs_rt(self, T)
+    .. method:: get_equilibrium_constants(self, T)
+    .. method:: get_temperature(self, H_or_E, T_guess, Y, do_energy=False)
+    .. method:: __init__(self, usr_np=numpy)
```
Fixed in (2b2c989)

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -327,123 +326,228 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+        Specify a user-defined NUMPY namespace as (*usr_np*) to the constructor
+        of a given mechanism thermochemistry class.
+
+        usr_np
```
Fixed in (2b2c989).

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -327,123 +326,228 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+    .. method:: get_species_gibbs_rt(self, T)
+    .. method:: get_equilibrium_constants(self, T)
+    .. method:: get_temperature(self, H_or_E, T_guess, Y, do_energy=False)
+    .. method:: __init__(self, usr_np=numpy)
```
meh.  for me usr_np is more descriptive.  what does bulk mean?

MTCam commented on .ci/generate-example-mech.py:
```diff
@@ -0,0 +1,11 @@
+import pyrometheus as pyro
+
+
+if __name__ == "__main__":
```
Fixed in (ba8e442).

MTCam commented on .ci/generate-example-mech.py:
```diff
@@ -0,0 +1,11 @@
+
+
+if __name__ == "__main__":
+    narg = len(sys.argv)
```
Fixed in (ba8e442).

MTCam commented on pyrometheus/__init__.py:
```diff
@@ -327,123 +326,228 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
-        RT  = self.gas_constant * T
-        return rho * RT / mmw
+def _pyro_norm(usr_np, argument, normord):
+    # Wrap norm for scalars
```
This one is a bit better I think (ec25fa3).

inducer commented:
Heads up! I've changed the target branch here to master.

inducer commented on pyrometheus/__init__.py:
```diff
@@ -183,17 +161,34 @@ def gen(c, t):
 
-# {{{ equilibrium constants
+def _zeros_like(argument):
+    return 0 * argument
```
```suggestion
    # FIXME: This mishandles NaNs.
    return 0 * argument
```

inducer commented on pyrometheus/__init__.py:
```diff
@@ -216,124 +211,353 @@ def equilibrium_constants_expr(sol: ct.Solution, react: ct.Reaction, gibbs_rt):
 
-    def get_mix_molecular_weight(self, Y):
-        return 1/np.dot( self.iwts, Y )
+    If the argument is a regular ole number, it just uses regular numpy.abs,
```
```suggestion
    If the argument is a regular ole number, it uses :func:`numpy.abs`,
```

inducer commented on pyrometheus/__init__.py:
```diff
@@ -216,124 +211,353 @@ def equilibrium_constants_expr(sol: ct.Solution, react: ct.Reaction, gibbs_rt):
-    def get_mix_molecular_weight(self, Y):
-        return 1/np.dot( self.iwts, Y )
+    If the argument is a regular ole number, it just uses regular numpy.abs,
+    otherwise use a linalg norm from regular numpy, or from the user's numpy
+    clone if one was specified.
```
```suggestion
    otherwise it uses ``usr_np.linalg.norm``.
```

inducer commented on pyrometheus/__init__.py:
```diff
@@ -216,124 +211,353 @@ def equilibrium_constants_expr(sol: ct.Solution, react: ct.Reaction, gibbs_rt):
+            to :class:`numpy.ndarray` and is used to hold all types of (potentialy
+            volumetric) "bulk data", such as temperature, pressure, mass fractions,
+            etc. This parameter defaults to *actual numpy*, so it can be ignored
+            unless it is needed by the user (e.g. for GPU processing).
```
```suggestion
            unless it is needed by the user (e.g. for purposes of 
            GPU processing or automatic differentiation).
```

inducer commented on pyrometheus/__init__.py:
```diff
@@ -342,37 +566,119 @@ def get_temperature(self, H_or_E, T_guess, Y, do_energy=False):
-            T_i  += dT
-            if np.abs( dT ) < tol:
-                T = T_i
+        ones = _pyro_zeros_like(enthalpy_or_energy) + 1.0
+        t_i = t_guess * ones
```
What does this do?

MTCam replied:
I intended to allow the call:
`get_temperature(.., tguess=300.0, ...)`,
where tguess is a single scalar value.

inducer commented on pyrometheus/__init__.py:
```diff
@@ -342,37 +566,119 @@ def get_temperature(self, H_or_E, T_guess, Y, do_energy=False):
-        return T
-
-""", strict_undefined=True)
+        return t_i
```
```suggestion
        raise RuntimeError("Temperature iteration failed to converge")
```

inducer commented on pyrometheus/__init__.py:
```diff
@@ -342,37 +566,119 @@ def get_temperature(self, H_or_E, T_guess, Y, do_energy=False):
+            dt = -f / j
+            t_i += dt
+            if _pyro_norm(self.usr_np, dt, np.inf) < tol:
                 break
```
```suggestion
                return t_i
```

inducer reviewed (approved):
Just a few things, good to go subject to these.

MTCam commented:
Seems like we are basically good-to-go for merging #7, #4 to main and address mass clipping as a separate issue? @ecisneros8, @inducer

ecisneros8 commented:
I'm ok with merging, but we should not forget array calculations with Numpy in #20. I started testing this weekend but had to stop and don't know if I can come back to it until Friday. Thoughts or comments?

MTCam commented:
> I'm ok with merging, but we should not forget array calculations with Numpy in #20. I started testing this weekend but had to stop and don't know if I can come back to it until Friday. Thoughts or comments?

We won't forget.  #20 is a follow-on PR that extends Pyro to work with numpy arrays.  It will not go away when #7 is merged.

inducer commented:
I say let's merge this.
